tools/ioflow/ioflow-read.c

```diff
@@ -0,0 +1,225 @@
+/* 
+ * ioflow-read.c        C source code for read tracer. 
+ *
+ *
+ * Copyright (c) Google LLC
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ *
+ * 25-Aug-2020   Haoning Chen   Created this.
+ */
+
+#include <uapi/linux/ptrace.h>
+#include <linux/blkdev.h>
+#include <linux/fs.h>
+
+#define IO_FLOW_READ 1
+// Use python string replacement to import common header code
+[IMPORT_COMM]
+
+BPF_HASH(map, u64, struct val_t);    // Map of syscall data, with pid as key
+BPF_HASH(rqmap, struct request *, struct rqval_t);    // Map of request info for aync request handling
+
+BPF_HISTOGRAM(hist_vfs);
+BPF_HISTOGRAM(hist_pgcache);
+BPF_HISTOGRAM(hist_readpg);
+BPF_HISTOGRAM(hist_ext4readpg);
+BPF_HISTOGRAM(hist_blk);
+BPF_HISTOGRAM(hist_split);
+BPF_HISTOGRAM(hist_merge);
+// Async request handling histograms
+BPF_HISTOGRAM(hist_rq_queue);
+BPF_HISTOGRAM(hist_rq_service);
+
+BPF_PERF_OUTPUT(syscall_events);
+BPF_PERF_OUTPUT(rq_events);
+
+
+int vfs_read_entry(struct pt_regs *ctx, struct file *file) {
+    // Filter asyc and direct IO
+    if (!(file->f_op->read_iter) || file->f_flags & O_DIRECT)
+        return 0;
+    
+    u64 pid = bpf_get_current_pid_tgid();
+    struct val_t val = {0};
+    comm_vfs_entry(&val, file);
+    map.update(&pid, &val);
+    return 0;
+}
+
+int page_cache_entry(struct pt_regs *ctx) {
+    u64 pid = bpf_get_current_pid_tgid();
+    struct val_t *valp = map.lookup(&pid);
+    if (valp && valp->ts_pgcache == 0)
+        valp->ts_pgcache = bpf_ktime_get_ns();
+    return 0;
+}
+
+int read_page_entry(struct pt_regs *ctx) {
+    u64 pid = bpf_get_current_pid_tgid();
+    struct val_t *valp = map.lookup(&pid);
+    if (valp && valp->ts_readpg <= valp->ts_pgcache)
```

Comment:
why check against ts_pgcache and not zero? were we getting cases where we were hitting read_page_entry before page_cache_entry etc.?

Reply:
I remember that initially I did check against 0, then I found there's a certain latter function (read_page_entry or write_page_entry, or ext4_read_page_entry, I can't remember exactly which one now) get hit ahead of its former functions, in some cases. 

But I just changed back to check against 0 and tested both read and write tracer. They both don't have problems. So I'll use 0 for all the 4 functions.